### PR TITLE
fix(query): fix view select filter can not push down to scan

### LINKED ISSUE #9882: bug: Scan all block when query condition has view table field
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.


### Version

0.8.160

### What's Wrong?

```
CREATE VIEW view_table AS SELECT 
a AS a,
b AS b,
try_cast(get(try_parse_json(xx),'xx') as varchar) AS c,
FROM xx;
```

SQL A:
```
MySQL [(none)]> explain select xx from view_table   where a = xx and b >= xx and b <= xx and c = 'xx';
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| explain                                                                                                                                                                                                                                                                                                                                   |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Exchange                                                                                                                                                                                                                                                                                                                                  |
| ├── exchange type: Merge                                                                                                                                                                                                                                                                                                                  |
| └── Filter                                                                                                                                                                                                                                                                                                                                |
|     ├── filters: [=(view_table.a (#0), xx), >=(view_table.b (#7), xx), <=(view_table.b (#7), xx), =(view_table.c (#25), xx)]                                                     |
|     └── EvalScalar                                                                                                                                                                                                                                                                                                                        |
|         ├── expressions: [CAST(xxx,CAST(get(try_parse_json(view_table.xx (#12)), xx) AS VARCHAR NULL)]       |
|         └── TableScan                                                                                                                                                                                                                                                                                                                     |
|             ├── table: view_table                                                                                                                                                                                                                                                                      |
|             ├── read rows: 1735586699                                                                                                                                                                                                                                                                                                     |
|             ├── read bytes: 3553019486915                                                                                                                                                                                                                                                                                                 |
|             ├── partitions total: 161416                                                                                                                                                                                                                                                                                                  |
|             ├── partitions scanned: 161416                                                                                                                                                                                                                                                                                                |
|             ├── push downs: [filters: [], limit: NONE]                                                                                                                                                                                                                                                                                    |
|             └── output columns: [0, 4, 6, 7, 8, 10, 12, 13, 21]                                                                                                                                                                                                                                                                           |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

SQL B:
```
MySQL [(none)]> explain select xx from view_table   where a = xx and b >= xx and b <= xx;
-------------------------------------------------------------------------------------+
| explain                                                                                                                                                                                                                                                                                                                               |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Exchange                                                                                                                                                                                                                                                                                                                              |
| ├── exchange type: Merge                                                                                                                                                                                                                                                                                                              |
| └── EvalScalar                                                                                                                                                                                                                                                                                                                        |
|     ├── expressions: [xx, CAST(get(try_parse_json(view_table.xx (#12)), xx) AS VARCHAR NULL)]       |
|     └── TableScan                                                                                                                                                                                                                                                                                                                     |
|         ├── table: view_table                                                                                                                                                                                                                                                                 |
|         ├── read rows: 8188707                                                                                                                                                                                                                                                                                                        |
|         ├── read bytes: 16898182568                                                                                                                                                                                                                                                                                                   |
|         ├── partitions total: 161419                                                                                                                                                                                                                                                                                                  |
|         ├── partitions scanned: 754                                                                                                                                                                                                                                                                                                   |
|         ├── push downs: [filters: [(a = xx), (b >= xx), (b <= xx)], limit: NONE]                                                                                                                                                                                             |
|         └── output columns: [0, 4, 6, 7, 8, 10, 12, 13, 21]                                                                                                                                                                                                                                                                           |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

SQL A just more than SQL B one column, but scanned all the blocks.


### How to Reproduce?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Replacing `DerivedColumn` in `Filter` with the column expression defined in the view, so that we can push down the `Filter` to the `Scan`.

Closes #9882

### CODE REVIEW

File: src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs [hunk 1]
```diff
@@ -59,6 +70,235 @@ impl RulePushDownFilterEvalScalar {
...
+                        if item.index == column.column.index {
+                            return item.scalar.clone();
+                        }
+                    }
+                    unreachable!()
+                }
+                ScalarExpr::BoundInternalColumnRef(_) | ScalarExpr::ConstantExpr(_) => {
+                    predicate.clone()
+                }
+                ScalarExpr::AndExpr(scalar) => {
```
Comment (sundy-li) [inline src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs:97]: could be after #10635

Comment (b41sh) [reply src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs:97]: ok

File: src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs [hunk 2]
```diff
@@ -59,6 +70,235 @@ impl RulePushDownFilterEvalScalar {
...
+            && predicate.used_columns().is_subset(eval_scalar_columns)
+        {
+            match predicate {
+                ScalarExpr::BoundColumnRef(column) => {
+                    for item in items {
+                        if item.index == column.column.index {
+                            return item.scalar.clone();
+                        }
+                    }
+                    unreachable!()
```
[the remark this answers is not in the document]
Comment (b41sh) [reply src/query/sql/src/planner/optimizer/rule/rewrite/rule_push_down_filter_eval_scalar.rs:92]: ok

**Review by xudong963 — APPROVED** [review]

**Comment by sundy-li** [comment]
cc @b41sh  Seems this pr is incomplete.

```
CREATE TABLE `table_a` (
  `source_a` BIGINT UNSIGNED NULL,
  `source_b` VARCHAR NULL
) ENGINE=FUSE;
CREATE VIEW `default`.`table_b` AS SELECT source_a AS c_1, TRY_CAST(get(try_parse_json(source_b), 'test') AS STRING) AS c_2 FROM table_a;
```

`c1` is not replaced into `source_a`

```
MySQL [(none)]> explain select * from table_b where c_1 >1679619600000 ;
+------------------------------------------------------------------------------------------------------------------------------+
| explain                                                                                                                      |
+------------------------------------------------------------------------------------------------------------------------------+
| EvalScalar                                                                                                                   |
| ├── expressions: [TRY_CAST(get(try_parse_json(table_a.source_b (#1)), "test") AS String NULL)]                               |
| ├── estimated rows: 0.00                                                                                                     |
| └── Filter                                                                                                                   |
|     ├── filters: [is_true(table_b.c_1 (#0) > 1679619600000)]                                                                 |
|     ├── estimated rows: 0.00                                                                                                 |
|     └── TableScan                                                                                                            |
|         ├── table: default.default.table_a                                                                                   |
|         ├── read rows: 30                                                                                                    |
|         ├── read bytes: 438                                                                                                  |
|         ├── partitions total: 3                                                                                              |
|         ├── partitions scanned: 3                                                                                            |
|         ├── pruning stats: [segments: <range pruning: 3 to 3>, blocks: <range pruning: 3 to 3, bloom pruning: 0 to 0>]       |
|         ├── push downs: [filters: [is_true(table_b.c_1 (#0) > 1679619600000)], limit: NONE]                                  |
|         └── estimated rows: 30.00                                                                                            |
+------------------------------------------------------------------------------------------------------------------------------+
15 rows in set (0.039 sec)

```

**Comment by b41sh** [comment]
> cc @b41sh Seems this pr is incomplete.
> 
> ```
> CREATE TABLE `table_a` (
>   `source_a` BIGINT UNSIGNED NULL,
>   `source_b` VARCHAR NULL
> ) ENGINE=FUSE;
> CREATE VIEW `default`.`table_b` AS SELECT source_a AS c_1, TRY_CAST(get(try_parse_json(source_b), 'test') AS STRING) AS c_2 FROM table_a;
> ```
We just replace the `DerivedColumn`, `table_a.source_a` and `table_b.c_1` have the same `column_index`, so there is no need to replace it.

**Pull request merged**